### PR TITLE
Add P2P protocol types and client

### PR DESCRIPTION
**Pull request by wanderingbort**
This PR adds the wire types for the antelope P2P protocol as well as a basic client that can send P2P messages and an automated heartbeat (if desired).  

Since there is no shared "socket" implementation between browsers and node environments, the client expects implementors to provide an implementation for the P2PProvider interface.  This interface matches the node Socket type. However, the antelope P2P protocol over a TCP connection can arrive in-order but fragmented.  There is a utility class provided that accepts/produces envelopes in the antelope P2P over TCP format and handles re-assembly. 

A node specific example looks like this :
```javascript
const socket = new Socket();
socket.connect(...);

const provider = new SimpleEnvelopeP2PProvider(socket);
const client = new P2PClient({provider});

client.on('message', (m:P2P.NetMessage) => {
   ...
});

client.send(P2P.HandshakeMessage.from({
   ...
});


```

### CODE REVIEW

File: test/p2p.ts
Comment (jnordberg) [inline test/p2p.ts:140]: Lots of Buffer

**Review by jnordberg — CHANGES_REQUESTED** [review]
First of all, this is great! Both functionality and implementation wise!

It breaks the browser tests however, to merge this we need to remove reliance on node.js primitives.

I marked all places I saw that uses them. We also need to fix the data loading for the browser tests, see my comment in tests/p2p.ts.

To run browser tests locally you can use `make browser-test`

**Comment by wanderingbort** [comment]
Thanks for the pointer to the browser tests @jnordberg they are green now 😄

**Comment by wanderingbort** [comment]
I see the prettier failures... on it

**Comment by jnordberg** [comment]
Still had some Buffer references left, took the liberty to fix them. This is good to merge once you merge in my PR to your PR :D https://github.com/wanderingbort/eosio-core/pull/1

**Comment by jnordberg** [comment]
Published in version 0.6.9

**Pull request merged**
